torch/distributed/_spmd/iter_graph_module.py

```diff
@@ -0,0 +1,393 @@
+
+        self.num_extra_output = 0
+
+    def _lookup_node(self, node: fx.Node, graph: fx.Graph) -> Optional[fx.Node]:
+        if graph == self.setup_graph:
+            return self._setup_mapping.get(node, None)
+        elif graph == self.cleanup_graph:
+            return self._cleanup_mapping.get(node, None)
+        return node
+
+    def _insert_context(self, func: str, node: fx.Node):
+        with ExitStack() as stack:
+            for graph in self._all_graphs:
+                if node:
+                    actual_node = self._lookup_node(node, graph)
+                    assert actual_node is not None, "Cannot handle None case now."
+                else:
+                    actual_node = node
+                stack.enter_context(getattr(graph, func)(actual_node))
+            yield
+
+    @contextmanager
+    def inserting_after(self, node):
+        return self._insert_context("inserting_after", node)
+
+    @contextmanager
+    def inserting_before(self, node):
+        return self._insert_context("inserting_before", node)
+
+    @staticmethod
+    def _is_sese_graph(nodes: List[fx.Node], graph: fx.Graph) -> bool:
+        """
+        Check if the given subgraph is a single-entry-single-exit (SESE)
+        graph of the original fx graph.
+        """
+        all_nodes: Set[fx.Node] = set(nodes)
+        for i, node in enumerate(nodes):
+            pytree_args, _ = tree_flatten(node.args)
+            pytree_kwargs, _ = tree_flatten(node.kwargs)
+            for arg in itertools.chain(pytree_args, pytree_kwargs):
+                if node not in all_nodes and i > 0:
+                    return False
+            if i == len(nodes) - 1:
+                # TODO: the only user must be the output. Otherwise, we don't
+                # know how to move this subgraph. We currently do not stricly
+                # force this attribute because some test code has orphan nodes.
+                if len(node.users) > 1:
+                    return False
+            else:
+                for user in node.users:
+                    if user not in all_nodes:
+                        return False
+        return True
+
+    def _convert_sese_input_to_output(
+        self, nodes: List[fx.Node], graph: fx.Graph, erase_node: bool
+    ) -> None:
+        for output in reversed(graph.nodes):
+            if output.target == "output":
+                break
+        first_node = self._lookup_node(nodes[0], graph)
+        assert first_node is not None, "The first_node is None."
+        # TODO: We currently assume there is only one input to simplify the
+        # coding. We may have to deal with a more general case.
+        sese_arguments = first_node.args[0]
+        # TODO: Do we need to remove the output of the SESE subgraph?
+        # new_output =  tuple(
+        #   arg for arg in output.args if arg != nodes[-1]
+        # ) + (arguments,)
+        new_output = output.args + (sese_arguments,)
```

Comment:
Isn't the first_node the input node to this graph? Can you clarify why the input and output of the graph is converted to be the new output? I would expect only the output to be converted to input.

Reply:
Since the entire subgraph will be moved, both the input and output of the subgraph will be modified. The input will from the `last iteration` which will be passed in through the input of the full graph. The output will become the input of the `next iteration` which will be carried over via the output of the full graph.